File: array/54.py
from typing import List
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if len(matrix) == 0: return []
        if len(matrix[0]) == 0: return []
        m = len(matrix) # 行
        n = len(matrix[0]) # 列
        
        dr_val = 0
        decay = [0 , 1, 1, 2]
        decay_base = 0

        ret = []
        i = 0
        j = -1
        while True:
            if dr_val == 0:
                if n - decay[dr_val] - decay_base == 0: return ret
                for _ in range(n - decay[dr_val] - decay_base):
                    j += 1
                    ret.append(matrix[i][j])
                dr_val = 1
            elif dr_val == 1:
                if m - decay[dr_val] - decay_base == 0: return ret
                for _ in range(m - decay[dr_val] - decay_base):
                    i += 1
                    ret.append(matrix[i][j])
                dr_val = 2
            elif dr_val == 2:
                if n - decay[dr_val] - decay_base == 0: return ret
                for _ in range(n - decay[dr_val] - decay_base):
                    j -= 1
                    ret.append(matrix[i][j])
                dr_val = 3
            elif dr_val == 3:
                if m - decay[dr_val] - decay_base == 0: return ret
                for _ in range(m - decay[dr_val] - decay_base):
                    i -= 1
                    ret.append(matrix[i][j])
                dr_val = 0
                decay_base += 2
        
def main():
    matrix = [
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9,10,11,12]
]
    print(Solution().spiralOrder(matrix))

if __name__ == "__main__":
    main()

File: leetcode-cpp/54.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        if (m == 0) return vector<int>();
        int n = matrix[0].size();
        int minus_inter = 0, dir = 0;
        int i = 0, j = -1;
        vector<int> ret;
        int decay_base[4] = {0, 1, 1, 2};
        while (true) {
            if (dir == 0) {
                int temp = n - decay_base[dir] - minus_inter;
                if (temp == 0) return ret;
                for (int t = 0; t < temp; t++){
                    j++;
                    ret.push_back(matrix[i][j]);
                }
                dir = 1;
            } else if (dir == 1) {
                int temp = m - decay_base[dir] - minus_inter;
                if (temp == 0) return ret;
                for (int t = 0; t < temp; t++){
                    i++;
                    ret.push_back(matrix[i][j]);
                }
                dir = 2;
            } else if (dir == 2) {
                int temp = n - decay_base[dir] - minus_inter;
                if (temp == 0) return ret;
                for (int t = 0; t < temp; t++){
                    j--;
                    ret.push_back(matrix[i][j]);
                }
                dir = 3;
            } else {
                int temp = m - decay_base[dir] - minus_inter;
                if (temp == 0) return ret;
                for (int t = 0; t < temp; t++){
                    i--;
                    ret.push_back(matrix[i][j]);
                }
                dir = 0;
                minus_inter += 2;
            }
        }
    }
};

int main(){
    Solution s;
    vector<vector<int>> matrix({{1,2,3},{4,5,6},{7,8,9}});
    vector<int> ret = s.spiralOrder(matrix);
    for (auto i : ret){
        cout <<i <<endl;
    }
}
